wire-runtime/src/main/java/com/squareup/wire/Message.java

```diff
@@ -149,12 +147,30 @@ protected Builder() {
      * this message has no unknown fields.
      */
     public final ByteString buildUnknownFields() {
-      return unknownFieldsBuffer != null
-          ? unknownFieldsBuffer.clone().readByteString()
-          : ByteString.EMPTY;
+      if (unknownFieldsBuffer != null) {
+        // Reads and caches the unknown fields from the buffer.
+        unknownFieldsByteString = unknownFieldsBuffer.readByteString();
+        unknownFieldsBuffer = null;
```

Comment:
I think that’s probably a bad fit in practice. It means handling unknown fields is O(N*N), which is a trivial DoS on our servers.